### ISSUE TITLE
updating/inserting/deleting comment()'s before the root element

### ISSUE THREAD

**Issue by mgr34** [eXist-db/exist#628]
> **notes**: 
>  1. *filling issue at the behest of Jörn from eXist solutions*
>  2. *[original mailing list submission][1]*

Not able to update (or insert) comments before the root element. Take for example

**db/test.xml**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!-- to be replaced -->
<document>
    <p>Contains stuff.</p>
</document>
````
**db/update.xq**
```xquery
xquery version "3.1";

let $doc := doc('/db/test.xml')
let $comment := $doc/comment()
return 
    update replace $comment with comment { "replaced" }
````

**Responds with:**

>exerr:ERROR The root element of a document can not be replaced with 'update replace'. Please consider removing the document or use 'update value' to just replace the children of the root. [at line 6, column 5, source: String]

**expected db/test.xml**:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!-- replaced -->
<document>
    <p>Contains stuff.</p>
</document>
````

I am currently working  develop branch  from Apr 15. (4a94ebd). On Debian 7.8. I suspect this is likely the case in eXist 2.2, but cannot confirm. 

[1]: https://sourceforge.net/p/exist/mailman/exist-open/thread/CAOFRqdgwTEZ6Np_kx-LwcWA5b13W2s_9q3cUYmi%2B7A92dzFAXw%40mail.gmail.com/#msg34107123

**Comment by joewiz** [comment]
Possibly related to #623?

**Comment by mgr34** [comment]
@joewiz I made that original stackoverflow post so it is related there. But I think otherwise this issue may be unrelated. 

For example if you look at the [particular error message displayed ][1] it looks at the `parentNode` and throws an error if its null, But in this case that makes sense. But comments and processing instructions would still be valid?  

```java
parent = (ElementImpl) node.getParentStoredNode();
 if (parent == null)
                    {throw new EXistException("The root element of a document can not be replaced with 'xu:replace'. " +
                        "Please consider removing the document or use 'xu:update' to just replace the children of the root.");}
```


My interpretation could be wrong. I didn't spend that much time actually trying to digest what was happening. 

[1]: https://github.com/eXist-db/exist/blob/30376d61ad9d81b2001001150c55b2beaa95c7b8/src/org/exist/xupdate/Replace.java#L74-L77

**Comment by joewiz** [comment]
After discussing this with @wolfgangmm in Oxford, I understand the XQuery Update code would need to be drastically reworked to allow update expressions to modify comment nodes outside the root element. I propose this be closed and labelled "won't fix."

**Comment by adamretter** [comment]
@joewiz I do think this could be fixed, but it would require some rearchitecting of the XQuery Update code. I  do consider this to be a bug, simply saying its a lot of work does not change that IMHO. We should leave the ticket open as it serves as a reference.

**Comment by joewiz** [comment]
@adamretter Fair enough. Given the choice between adding this feature and adding official XQuery Update Facility support to eXist, what would you say is the best use of resources?

**Comment by joewiz** [comment]
(Implicit in my statement was that XQUF would work on persistent as well as in-memory nodes.)

**Comment by adamretter** [comment]
XQUF is the way to go, maintaining something non-standard and broken in bad. However, adding XQUF is much more work of coursevand so their is a quality vs time/resources trade off.

Also if people wanted to apply XQUF inline, then we would need some parts of XQuery Scripting too.
